### PR TITLE
API: added partial bitstream output API

### PR DESCRIPTION
**Pull request by pvf-intel**

### CODE REVIEW

File: doc/mediasdk-man.md [hunk 1]
```diff
@@ -7564,6 +7564,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+    mfxExtBuffer    Header;
+    mfxU32          BlockSize;        /* output block granulatiry for Granularity = MFX_PARTIAL_BITSTREAM_BLOCK */
+    mfxU16          Granularity;      /* granulatiry of the partial bitstream: slice/block/any */
+    mfxU16          reserved[8];
+} mfxExtPartialBitstreamParam;
+```
+
+**Description**
+
+This structure is used by AVC HW encoder to output parts of bitstream as soon as they ready. The application can attach this extended buffer to the [mfxFrameSurface1](#mfxFrameSurface) structure at init time.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: > "to the mfxFrameSurface1 structure at init time"

Confusing...

How "init time" for mfxFrameSurface1 is defined? There is no corresponding dedicated API call. Do you mean Encoding Init() function.

File: doc/mediasdk-man.md [hunk 2]
```diff
@@ -7564,6 +7564,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+If this option is turned ON and session configured in single thread, then AVC encoder can output bitstream by part based with required granularity. <br>
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: > option is turned ON

In this context (descring ext buffer structure) is confusing. What means "turned ON" - when buffer is attached to frame surface or anything else?

> session configured in single thread

What do you mean here? AsyncDepth=1 or mediasdk library initialization via MFXInitEx with threads=1?

File: doc/mediasdk-man.md [hunk 3]
```diff
@@ -7564,6 +7564,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+<br>This parameter is valid only during initialization and reset. Absence of this buffer means default bitstream output behavior.<br>
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: During initialization and reset I have mfxVideoParam and I don't have mfxFrameSurface1...

File: doc/mediasdk-man.md [hunk 4]
```diff
@@ -7564,6 +7564,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+<br>Not all codecs and SDK implementations support this value. Use [Query](#MFXVideoENCODE_Query) function to check if this feature is supported.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: > this value

"feature" I guess

File: doc/mediasdk-man.md [hunk 5]
```diff
@@ -8293,6 +8330,8 @@ SDK API 1.6 added the `MFX_WRN_FILTER_SKIPPED` return status.
 
 SDK API 1.19 added `MFX_ERR_GPU_HANG` and `MFX_ERR_REALLOC_SURFACE`.
 
+SDK API 1.31 added `MFX_ERR_NONE_PARTIAL_OUTPUT`
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: 1.31 or TBD? If we miss something in this repository from API 1.31, this should be added in a separate commit.

File: doc/mediasdk-man.md [hunk 6]
```diff
@@ -8994,6 +9033,26 @@ The `Protected` enumerator describes the protection schemes.
...
+The `PartialBitstreamOutput` enumerator indicates flags of partial bitstream output type. 
+
+**Name/Description**
+
+| | |
+--- | ---
+`MFX_PARTIAL_BITSTREAM_NONE`  | Don't use partial output
+`MFX_PARTIAL_BITSTREAM_SLICE` | Partial bitstream output will be aligned to slice granularity (AVC) 
+`MFX_PARTIAL_BITSTREAM_BLOCK` | Partial bitstream output will be aligned to user-defined block size granularity
+`MFX_PARTIAL_BITSTREAM_ANY` | Partial bitstream output will be return any coded data avilable at the end of SyncOperation timeout
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: It sounds that "MFX_PARTUAL_BITSTREAM_FULL" will be better since it highlights that all data should be returned, i.e. no partial

File: builder/ConfTargets.cmake [hunk 7]
```diff
@@ -106,6 +106,7 @@ cmake_dependent_option(
 option( MFX_ENABLE_VP9_VIDEO_ENCODE "Enable VP9 encoder?" ON)
 
 option( MFX_ENABLE_ASC "Enable ASC support?"  ON )
+option( MFX_ENABLE_LOW_DELAY_BS_OUT "Enable Partial Bitstream optput support?"  OFF )
```
Comment (dvrogozh) [inline builder/ConfTargets.cmake]: Should we really have dedicated command line option for that? You see, cmdline cmake options were introduced to switch on/off big features like entire codec or enable/disable EU shaders. This particular new feature does not seem to imply any new EU shader, does it? Can you, please, hint how it will be implemented? If it will be handled on mediasdk .cpp level and code won't be dramatically big, I don't think we need a special switch for it on cmdline level, just ever-on define in config.h will be enough.

File: api/include/mfxdefs.h [hunk 8]
```diff
@@ -181,6 +181,11 @@ typedef enum
     MFX_WRN_FILTER_SKIPPED              = 10,   /* one of requested filters has been skipped */
     MFX_WRN_INCOMPATIBLE_AUDIO_PARAM    = 11,   /* incompatible audio parameters */
 
+#if MFX_VERSION >= MFX_VERSION_NEXT
+    /* low-delay partial output */
+    MFX_ERR_NONE_PARTIAL_OUTPUT         = 12,   /* frame is not ready, but bitstream contains partial output */
```
Comment (dmitryermilov) [inline api/include/mfxdefs.h]: As far as I remember we agreed not to add such status because application won't use it. Please explain what has changed?

Comment (pvf-intel) [reply api/include/mfxdefs.h]: This one is required as we need to report status of incomplete bitstream in sync operation. You confuse it with internal status to respin scheduler.

File: doc/mediasdk-man.md (hunk 1 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: You right it should be mfxVideoParam instead of mfxFrameSurface1

File: doc/mediasdk-man.md (hunk 2 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: "turned ON" means Granularity !=MFX_PARTIAL_BITSTREAM_NONE and buffer is attached
"session configured in single thread" means mfxIMPL in the session initialization should contain MFX_IMPL_EXTERNAL_THREADING

File: doc/mediasdk-man.md (hunk 3 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: See comment before, you right

File: doc/mediasdk-man.md (hunk 4 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: right, will be fixed

File: doc/mediasdk-man.md (hunk 6 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: I am not agree. FULL means full bitstream while ANY means any incomplete data available upon sync operation call

File: builder/ConfTargets.cmake (hunk 7 above)
Comment (pvf-intel) [reply builder/ConfTargets.cmake]: Agree, will remove this option

File: doc/mediasdk-man.md (hunk 5 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: Is it right way to use TBD? Will it be changed to the version of next release?

File: api/include/mfxdefs.h (hunk 8 above)
Comment (dmitryermilov) [reply api/include/mfxdefs.h]: Okay. Thanks. So if application indeed can receive it, would you mind describing in the man how app should handle it?

Comment (pvf-intel) [reply api/include/mfxdefs.h]: It should rerun sync operation again until it return error or MFX_ERR_NONE

Comment (dmitryermilov) [reply api/include/mfxdefs.h]: in the man..

File: builder/ConfTargets.cmake (hunk 7 above)
Comment (dvrogozh) [reply builder/ConfTargets.cmake]: As of now it is not removed.

File: doc/mediasdk-man.md (hunk 1 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: No change in PR. unresolved.

File: doc/mediasdk-man.md (hunk 4 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: not yet done

File: doc/mediasdk-man.md (hunk 5 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: Yes, it should be changed by @onabiull

File: doc/mediasdk-man.md (hunk 6 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: I misread the sentence. Please, fix typos to make it readable:
* s/will be return/will return/
* s/avilable/available/

> at the end of SyncOperation timeout

I think this requires explicit comment in the man. Do I understand correctly that if I will call SyncOperation(timeout) then on the point of exit from SyncOperation with WRN_IN_EXECUTION I will get some data in the buffer? If yes, do you have example in the manual how to use that? Which timeout to specify?

File: doc/mediasdk-man.md (hunk 2 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: I would add in the man: "If this option is turned ON, i.e. mfxExtPartialBitstreamParam is attached and Granularity !=MFX_PARTIAL_BITSTREAM_NONE, ..." to make it clear.

I don't see MFX_IMPL_EXTERNAL_THREADING description in mediasdk manual. What is that?

**Comment by dvrogozh** [comment]
Oh my God. Please, squash commits. There is no point to be iterative to this degree. For me it obscures the change and I can't review thoroughly.

File: api/include/mfxdefs.h (hunk 8 above)
Comment (aleonenk) [reply api/include/mfxdefs.h]: Agree. Pavel, please add a section to the SyncOperation documentation, describing the workflow around MFX_ERR_NONE_PARTIAL_OUTPUT return code, including some information about bitstream buffer management

File: api/include/mfxstructures.h [hunk 9]
```diff
@@ -124,6 +124,9 @@ enum {
     MFX_FOURCC_Y216         = MFX_MAKEFOURCC('Y','2','1','6'),
     MFX_FOURCC_Y416         = MFX_MAKEFOURCC('Y','4','1','6'),
 #endif
+#if (MFX_VERSION >= MFX_VERSION_NEXT)
+    MFX_EXTBUFF_PARTIAL_BITSTREAM_PARAM         = MFX_MAKEFOURCC('P','B','O','P'),
```
Comment (dvrogozh) [inline api/include/mfxstructures.h]: Why do you place extbuf param in the list of FOURCC color formats? If it is color format - name it accordingly. If it is not - place elsewhere.

File: doc/mediasdk-man.md [hunk 10]
```diff
@@ -1973,17 +1973,23 @@ This function initiates execution of an asynchronous function not already starte
...
 **Change History**
 
 This function is available since SDK API 1.0.
+Return code MFX_ERR_NONE_PARTIAL_OUTPUT is introduced in SDK API TBD.
 
 **Remarks**
 
 See status codes for specific asynchronous functions.
 
+In partial output mode, enabled during encoder initialization [mfxExtPartialBitstreamParam](#mfxExtPartialBitstreamParam), application is expected to call SyncOperation repeatedly until it receives the MFX_ERR_NONE (or some error);
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: repeatedly with which sync point, the same? I think it worths to be said explicitly in the man.

If library changes mfxBitstream in async part, then there will be race condition between library async part and end-user application. How user is supposed to handle it having no mutex to lock?

File: doc/mediasdk-man.md [hunk 11]
```diff
@@ -1973,17 +1973,23 @@ This function initiates execution of an asynchronous function not already starte
...
+when MFX_ERR_NONE_PARTIAL_OUTPUT code is returned, new portion of the encoded frame is located at DataOffset location and have a DataLength size in the associated [mfxBitstream](#mfxBitstream) buffer.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: This blows out my mind. So, I called SyncOperation for the first time. And while I am accessing first portion library might change DataOffset to point to the next portion? The behavior you describe is direct description of the race condition. Please, clarify/change. Also, please, provide code example right in the manual.

File: doc/mediasdk-man.md [hunk 12]
```diff
@@ -7564,6 +7570,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+    mfxU32          BlockSize;        /* output block granulatiry for Granularity = MFX_PARTIAL_BITSTREAM_BLOCK */
+    mfxU16          Granularity;      /* granulatiry of the partial bitstream: slice/block/any */
+    mfxU16          reserved[8];
+} mfxExtPartialBitstreamParam;
+```
+
+**Description**
+
+This structure is used by AVC HW encoder to output parts of bitstream as soon as they ready. The application can attach this extended buffer to the [mfxVideoParam](#mfxVideoParam) structure at init time.
+If this option is turned ON (Granularity != MFX_PARTIAL_BITSTREAM_NONE) and session configured in single thread (session initialized with MFX_IMPL_EXTERNAL_THREADING), then AVC encoder can output bitstream by part based with required granularity. <br>
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: You refer to something which is not described in the documentation, but exists in the header file and in mediasdk dispatcher:
```
[dvrogozh@dvrscl msdk]$ fgrep -rsn MFX_IMPL_EXTERNAL_THREADING
api/include/mfxcommon.h:67:    MFX_IMPL_EXTERNAL_THREADING        = 0x10000,
api/mfx_dispatch/windows/src/main.cpp:68:        {MFX_LIB_HARDWARE, MFX_IMPL_HARDWARE|MFX_IMPL_EXTERNAL_THREADING, 0},
api/mfx_dispatch/windows/src/main.cpp:69:        {MFX_LIB_HARDWARE, MFX_IMPL_HARDWARE2|MFX_IMPL_EXTERNAL_THREADING, 1},
api/mfx_dispatch/windows/src/main.cpp:70:        {MFX_LIB_HARDWARE, MFX_IMPL_HARDWARE3|MFX_IMPL_EXTERNAL_THREADING, 2},
api/mfx_dispatch/windows/src/main.cpp:71:        {MFX_LIB_HARDWARE, MFX_IMPL_HARDWARE4|MFX_IMPL_EXTERNAL_THREADING, 3},
api/mfx_dispatch/windows/src/mfx_library_iterator.cpp:66:    mfxIMPL impl_via = (*pImplInterface & ~MFX_IMPL_EXTERNAL_THREADING);
```

This gives a bunch of questions what is this mode, is it really functional and is it really implemented?

File: mfxconfig.h.in [hunk 13]
```diff
@@ -23,6 +23,7 @@
 #cmakedefine MFX_ENABLE_MCTF
 #cmakedefine MFX_ENABLE_ASC
 #cmakedefine MFX_ENABLE_CPLIB
+#cmakedefine MFX_ENABLE_LOW_DELAY_BS_OUT
```
Comment (dvrogozh) [inline mfxconfig.h.in]: please, delete.

File: _studio/shared/include/mfx_config.h [hunk 14]
```diff
@@ -113,6 +113,12 @@
     #define MFX_ENABLE_FOURCC_RGB565
 #endif
 
+#if defined(MFX_ENABLE_LOW_DELAY_BS_OUT)
```
Comment (dvrogozh) [inline _studio/shared/include/mfx_config.h]: delete (enable macro)

File: builder/profiles/fastboot.cmake [hunk 15]
```diff
@@ -33,6 +33,7 @@ set( MFX_ENABLE_KERNELS OFF )
 set( MFX_ENABLE_SW_FALLBACK OFF )
 set( MFX_ENABLE_MCTF OFF )
 set( MFX_ENABLE_ASC OFF )
+set( MFX_ENABLE_LOW_DELAY_BS_OUT OFF )
```
Comment (dvrogozh) [inline builder/profiles/fastboot.cmake]: delete

File: doc/mediasdk-man.md (hunk 10 above)
Comment (aleonenk) [reply doc/mediasdk-man.md]: - We reuse the same sync point.
- Library update the app provided bitstream structure only during the SyncOperation call, this should be explicitly documented (any processing of the partial output on app level must be finished before next SyncOperation call). Good catch.

File: _studio/shared/include/mfx_config.h (hunk 14 above)
Comment (pvf-intel) [reply _studio/shared/include/mfx_config.h]: Whitch macro do you mean? Where should it be added?

File: doc/mediasdk-man.md (hunk 10 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: I guess the same possibility for race condition is true for any other bitstream changes in library and end-user application. Nothing special here.

File: doc/mediasdk-man.md (hunk 11 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: As commented before. There is nothing special for this feature. The same behavior can happen in normal operation mode when SyncOperation returns IN_EXECUTION status.

File: doc/mediasdk-man.md (hunk 10 above)
Comment (dvrogozh) [reply doc/mediasdk-man.md]: No. If SyncOperation returned MFX_ERR_NONE then library is done with the bitsteam and won't modify it. I guess that what you really want to say is that partial bitstream handling will work in a very special mode only (external threading?), i.e. when there is synchronous processing and library can't actually change the bitstream while user works with it. I think that's what needs to be clarified.

File: _studio/shared/include/mfx_config.h (hunk 14 above)
Comment (dvrogozh) [reply _studio/shared/include/mfx_config.h]: I think that MFX_ENABLE_LOW_DELAY_BS_OUT should be removed from the code.

File: doc/mediasdk-man.md (hunk 10 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: But if it returns MFX_WRN_IN_EXECUTION then SynOperation is incomplete and there is possibility to change bitstream simultaneously in library and from end-user side from different threads.

Comment (pvf-intel) [reply doc/mediasdk-man.md]: I guess also if end-users uses one mfxBistream for multiple frames sinking data after each frame it is possible to have race conditions because bitstream modification happens on event from hw (or scheduler in single thread mode) that can happen during end-user bitstream modification.

File: doc/mediasdk-man.md (hunk 4 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: done

File: doc/mediasdk-man.md (hunk 5 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: done

File: doc/mediasdk-man.md (hunk 6 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: done

File: builder/ConfTargets.cmake (hunk 7 above)
Comment (pvf-intel) [reply builder/ConfTargets.cmake]: done

File: doc/mediasdk-man.md [hunk 16]
```diff
@@ -2187,17 +2187,25 @@ This function initiates execution of an asynchronous function not already starte
...
+Return code MFX_ERR_NONE_PARTIAL_OUTPUT is introduced in SDK API TBD.
 
 **Remarks**
 
 See status codes for specific asynchronous functions.
 
+In partial output mode, enabled during encoder initialization [mfxExtPartialBitstreamParam](#mfxExtPartialBitstreamParam) or reset, application is expected to call SyncOperation with the same sync point repeatedly until it receives the MFX_ERR_NONE or some error;
+when MFX_ERR_NONE_PARTIAL_OUTPUT code is returned, new portion of the encoded frame is appended to the last bitstream portion located at DataOffset and increment a DataLength size accoding to new portion size in the associated [mfxBitstream](#mfxBitstream) buffer.
+Finally, when MFX_ERR_NONE returned the bitstream buffer contains full bitstream for encoded frame with DataLengh equal to it length. 
+Any processing on app level of the partial result returned in bitstream buffer must be finished before next SyncOperation call.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: Just to double check: we guarantee that mediasdk will change mfxBitstream description (DataLength at the first place) only within SyncOperation call?

File: doc/mediasdk-man.md [hunk 17]
```diff
@@ -7872,6 +7880,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+    mfxU16          reserved[8];
+} mfxExtPartialBitstreamParam;
+```
+
+**Description**
+
+This structure is used by AVC HW encoder to output parts of bitstream as soon as they ready. The application can attach this extended buffer to the [mfxVideoParam](#mfxVideoParam) structure at init time.
+If this option is turned ON (Granularity != MFX_PARTIAL_BITSTREAM_NONE), then AVC encoder can output bitstream by part based with required granularity. <br>
+<br>This parameter is valid only during initialization and reset. Absence of this buffer means default or previously configured bitstream output behavior.<br>
+<br>Not all codecs and SDK implementations support this feature. Use [Query](#MFXVideoENCODE_Query) function to check if this feature is supported.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: In above text you directly claim which codec supports this feature: AVC HW! I think you need to remove AVC HW from the above text.

File: doc/mediasdk-man.md [hunk 18]
```diff
@@ -7872,6 +7880,37 @@ This structure is used to pass decryption status report index for Common Encrypt
...
+    mfxExtBuffer    Header;
+    mfxU32          BlockSize;        /* output block granulatiry for Granularity = MFX_PARTIAL_BITSTREAM_BLOCK */
+    mfxU16          Granularity;      /* granulatiry of the partial bitstream: slice/block/any */
+    mfxU16          reserved[8];
+} mfxExtPartialBitstreamParam;
+```
+
+**Description**
+
+This structure is used by AVC HW encoder to output parts of bitstream as soon as they ready. The application can attach this extended buffer to the [mfxVideoParam](#mfxVideoParam) structure at init time.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: > by AVC HW encoder

Actually structure might apply to any codec. Moreover, below you suggest to Query and check whether codec supports the feature or not. I think it is better to remove any mentioning of the specific codec (AVC) from this text.

File: doc/mediasdk-man.md [hunk 19]
```diff
@@ -8447,6 +8486,8 @@ If the acceleration device is not default device, use the following values to in
 `MFX_IMPL_HARDWARE2` | Use the hardware-accelerated implementation on the second acceleration device.
 `MFX_IMPL_HARDWARE3` | Use the hardware-accelerated implementation on the third acceleration device.
 `MFX_IMPL_HARDWARE4` | Use the hardware-accelerated implementation on the fourth acceleration device.
+`MFX_IMPL_SINGLE_THREAD` |  Prohibits creation of internal threads. Do not create its own threads in MediaSDK library and plugin and work only in application threads.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: Can you, please move this out in a separate commit preceding current one? That's separate feature from partial bitstream output.

File: doc/mediasdk-man.md (hunk 16 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: yes, it works in the same way currently for frame based encoding, no difference for partial output here

File: doc/mediasdk-man.md (hunk 19 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: fixed in new update of pull request

File: doc/mediasdk-man.md (hunk 18 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: fixed in new update of pull request

File: doc/mediasdk-man.md (hunk 17 above)
Comment (pvf-intel) [reply doc/mediasdk-man.md]: fixed in new update of pull request

File: doc/mediasdk-man.md [hunk 20]
```diff
@@ -9338,6 +9385,25 @@ The `mfxComponentType` enumerator describes type of workload passed to [MFXQuery
...
+**Description**
+
+The `PartialBitstreamOutput` enumerator indicates flags of partial bitstream output type. 
+
+**Name/Description**
+
+| | |
+--- | ---
+`MFX_PARTIAL_BITSTREAM_NONE`  | Don't use partial output
+`MFX_PARTIAL_BITSTREAM_SLICE` | Partial bitstream output will be aligned to slice granularity (AVC)
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: Is it only applicable for AVC?

Comment (pvf-intel) [reply doc/mediasdk-man.md]: no, fixed

File: api/include/mfxdefs.h (hunk 8 above)
Comment (pvf-intel) [reply api/include/mfxdefs.h]: The section was added long time ago. Any update for review?

File: _studio/shared/src/mfx_static_assert_structs.cpp [hunk 21]
```diff
@@ -3612,3 +3612,9 @@
         MSDK_STATIC_ASSERT_STRUCT_OFFSET(mfxExtCam3DLut                     ,Table                         ,32   )
     #endif
 #endif //defined (__MFXCAMERA_H__)
+
+#if (MFX_VERSION >= MFX_VERSION_NEXT)
+        MSDK_STATIC_ASSERT_STRUCT_OFFSET(mfxExtPartialBitstreamParam        ,Header                        ,0  )
```
Comment (mgonchar) [inline _studio/shared/src/mfx_static_assert_structs.cpp]: Please also add assert on size of the ```mfxExtPartialBitstreamParam``` struct, these asserts starts from line 50 of this file
